Python/chapter03/p02_stack_min/miguelHx.py

```diff
@@ -0,0 +1,222 @@
+            int: min value in stack
+        """
+        return self.top.running_min
+
+    def peek(self) -> int:
+        """
+        Returns data at the top of the stack
+
+        Raises:
+            IndexError: [description]
+
+        Returns:
+            int: the value at the top of the stack
+        """
+        if self.top is None:
+            raise IndexError('Stack is Empty')
+        return self.top.data
+    
+    def is_empty(self) -> bool:
+        """
+        Returns True if and only if the
+        stack is empty, otherwise False.
+
+        Returns:
+            bool: True if empty, False otherwise
+        """
+        return self.size == 0
+    
+    def as_list(self) -> List[int]:
+        """
+        Builds a list of the current stack state.
+        For example, given the following stack:
+            3 -> 2 -> 1, where 3 is the top,
+        Expect:
+            [3, 2, 1]
+        Returns:
+            List[int]: list of integers
+        """
+        values = []
+        n = self.top
+        while n.next is not None:
+            values.append(n.data)
+            n = n.next
+        values.append(n.data)
+        return values
+    
+    def __str__(self):
+        if self.size == 0:
+            return '<Empty>'
+        values = []
+        n = self.top
+        while n.next is not None:
+            values.append(str(n.data))
+            n = n.next
+        values.append(str(n.data))
+        return '->'.join(values)      
+
+
+class TestMyStack(unittest.TestCase):
+    def test_stack_push(self):
+        s = MyStack()
+        self.assertEqual(s.size, 0)
+        self.assertEqual(s.top, None)
+        s.push(2)
+        self.assertEqual(s.size, 1)
+        self.assertEqual(s.top.data, 2)
+        self.assertEqual(s.top.next, None)
+        s.push(3)
+        self.assertEqual(s.size, 2)
+        self.assertEqual(s.top.data, 3)
+        self.assertEqual(s.top.next.data, 2)
+        s.push(4)
+        self.assertEqual(s.size, 3)
+        self.assertEqual(s.top.data, 4)
+        self.assertEqual(s.top.next.data, 3)
+
+        l = s.as_list()
+        self.assertEqual(l, [4, 3, 2])
+    
+    def test_stack_peek(self):
+        s = MyStack()
+        with self.assertRaises(IndexError):
+            s.peek()
+        s.push(1)
+        s.push(2)
+        s.push(99)
+        top_val = s.peek()
+        self.assertEqual(top_val, 99)
+    
+    def test_stack_is_empty(self):
+        s = MyStack()
+        self.assertTrue(s.is_empty())
+        s.push(7)
+        self.assertFalse(s.is_empty())
+    
+    def test_stack_pop(self):
+        # first case, attempt to pop an empty stack
+        s = MyStack()
+        with self.assertRaises(IndexError):
+            s.pop()
+        s.push(1)
+        s.push(2)
+        s.push(3)
+        # size is 3
+        self.assertEqual(s.as_list(), [3, 2, 1])
+        val = s.pop()
+        self.assertEqual(val, 3)
+        self.assertEqual(s.size, 2) # size should now be 2
```

Comment:
optional: it might be better to make `MyStack` implement [`__len__`](https://docs.python.org/3/reference/datamodel.html#object.__len__) and rename `size` to `_size`: that way callers of the class can't mess around with internals.

Reply:
Good idea, resolved in 745f6a4e1b08a36c0808435afd3c4bec186e8b5e